skimage/morphology/_flood_fill.py

```diff
@@ -0,0 +1,290 @@
+    """
+    # Correct start point in ravelled image - only copy if non-contiguous
+    image = np.asarray(image)
+    if image.flags.f_contiguous is True:
+        order = 'F'
+    elif image.flags.c_contiguous is True:
+        order = 'C'
+    else:
+        warnings.warn('Non-contiguous array passed as `image`; this will be '
+                      'converted to a contiguous array as a copy.')
+        image = np.ascontiguousarray(image)
+        order = 'C'
+
+    seed_value = image[seed_point]
+
+    # Shortcut for rank zero
+    if image.size == 0:
+        return np.array([], dtype=np.uint8)
+
+    # Convenience for 1d input
+    try:
+        iter(seed_point)
+    except TypeError:
+        seed_point = (seed_point,)
+
+    selem = _resolve_neighborhood(selem, connectivity, image.ndim)
+
+    # Must annotate borders
+    working_image = _fast_pad(image, image.min())
+
+    # Stride-aware neighbors - works for both C- and Fortran-contiguity
+    ravelled_seed_idx = np.ravel_multi_index([i+1 for i in seed_point],
+                                             working_image.shape, order=order)
+    neighbor_offsets = _offsets_to_raveled_neighbors(
+        working_image.shape, selem, center=((1,) * image.ndim))
+
+    # Use a set of flags; see _flood_fill_cy.pyx for meanings
+    flags = np.zeros(working_image.shape, dtype=np.uint8)
+    _set_edge_values_inplace(flags, value=2)
+
+    try:
+        if tolerance is not None:
+            # Check if tolerance could create overflow problems
+            try:
+                max_value = np.finfo(working_image.dtype).max
+                min_value = np.finfo(working_image.dtype).min
+            except ValueError:
+                max_value = np.iinfo(working_image.dtype).max
+                min_value = np.iinfo(working_image.dtype).min
+
+            high_tol = min(max_value, seed_value + tolerance)
+            low_tol = max(min_value, seed_value - tolerance)
+
+            _flood_fill_tolerance(working_image.ravel(),
+                                  flags.ravel(),
+                                  neighbor_offsets,
+                                  ravelled_seed_idx,
+                                  seed_value,
+                                  low_tol,
+                                  high_tol)
+        else:
+            _flood_fill_equal(working_image.ravel(),
+                              flags.ravel(),
+                              neighbor_offsets,
+                              ravelled_seed_idx,
+                              seed_value)
+    except TypeError:
+        if working_image.dtype == np.float16:
+            # Provide the user with clearer error message
+            raise TypeError("dtype of `image` is float16 which is not "
+                            "supported, try upcasting to float32")
+        else:
+            raise
+
+    # Output what the user requested; view does not create a new copy.
+    return flags[(slice(1, -1),) * image.ndim].view(np.bool)
```

Comment:
Seeing this: is there a convention in scikit-image on how functions return indices? E.g. as a boolean array of the same shape, as an array of 0s and 1s or an `np.nonzero(...)` of the former (see earlier comment)?

Reply:
Not currently. ;) I think we should return indices differently depending on the amount of work required. Since `flood` produces a boolean array as a side effect, I think we should return that, and not do extra work to get `np.nonzero` when users might not require it. Similarly, if something produced indices (`np.nonzero`-style) as a side effect, I don't think we should instantiate an enormous bool array.

Having said that, a counter-point: we can imagine a future in which NumPy arrays are expandable in amortised-constant time (similar to Python lists), in which case, the indices approach becomes much more natural, and we might want to switch to that. Do we plan for said future by normalising our API to the most "efficient" output perspective? I would argue not (who knows if expandable arrays will ever be a thing), but it's worth considering.

At any rate, I'll go back to what I said above: this should not hold up this PR.

Reply:
> I think we should return indices differently depending on the amount of work required. Since `flood` produces a boolean array as a side effect, I think we should return that, [...]

Well it doesn't satisfy the perfectionist inside me but it's probably the more pragmatic approach. I think it would be worth it though to decide between returning a boolean array or an array of 0s and 1s for functions that don't use indices. Creating a view is really cheap and always returning the same type would make the user experience more consistent. And by that I don't mean that we go around changing stuff but just that we have a guideline for future changes.

> At any rate, I'll go back to what I said above: this should not hold up this PR.

Fully agreed.

Reply:
> decide between returning a boolean array or an array of 0s and 1s for functions that don't use indices.

Definitely bool. 0s and 1s can't be used to index a NumPy array. (At least not to get the expected result!)

Reply:
Should we change that in `local_maxima` as well then?

Reply:
Holy crap! Yes! LOL

Reply:
The API breakage will be fun... =\

Reply:
As @jni said, this is basically due to minimizing computation time required to arrive at a result which meets NumPy requirements for indexing.  I think NumPy tightened their requirements in the last year or so to a Boolean array from looser prior tolerances.  I think I'd prefer a nested Tuple as well, but that's not realistic given how this algorithm works.